taichi/program/function_key.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <functional>
+#include <string>
+
+namespace taichi {
+namespace lang {
+
+class FunctionKey {
+ public:
+  std::string func_name;
+  int func_id;
```

Comment:
Explain that:

1. `func_id` maps to a `@ti.func`
2. `instance_id` is because a `@ti.func` may be instantiated multiple times
3. `func_name` is mostly for debugging/visualizing purpose, and doesn't need to participate in the hash computation?